### PR TITLE
Port to dune

### PR DESCRIPTION
**Pull request by TheLortex**
In the global effort of switching mirage to dune: mirage/mirage#969
Fixes https://github.com/mirage/mirage-solo5/issues/41

This is a work in progress, I'm trying to keep track of every change that need to be made!

### CODE REVIEW

File: lib/main.ml [hunk 1]
```diff
@@ -23,6 +23,8 @@
 
 open Lwt
 
+[@@@warning "-3"]
```
Comment (hannesm) [inline lib/main.ml]: why is this needed?

File: lib/time.ml [hunk 2]
```diff
@@ -2,9 +2,9 @@
  * http://www.ocsigen.org/lwt
  * Module Lwt_mirage, based on Lwt_unix
  * Copyright (C) 2010 Anil Madhavapeddy
- * Copyright (C) 2005-2008 J�r�me Vouillon
- * Laboratoire PPS - CNRS Universit� Paris Diderot
- *                    2009 J�r�mie Dimino
+ * Copyright (C) 2005-2008 J�r�me Vouillon
+ * Laboratoire PPS - CNRS Universit� Paris Diderot
+ *                    2009 J�r�mie Dimino
```
Comment (hannesm) [inline lib/time.ml]: seems to be an unintentional change (encoding issue)

File: lib/time.ml [hunk 3]
```diff
@@ -24,6 +24,7 @@
  *)
 
 open Lwt
+[@@@warning "-3"]
```
Comment (hannesm) [inline lib/time.ml]: see above, imho we should fix warnings instead of disabling them module-globally

File: lib/dune [hunk 4]
```diff
@@ -0,0 +1,18 @@
+(library
+  (name oS)
+  (public_name mirage-solo5.internals)
+  (modules lifecycle main mM oS solo5 time)
+  (c_names alloc_pages_stubs atomic_stubs barrier_stubs checksum_stubs clock_stubs cstruct_stubs main mm_stubs solo5_block_stubs solo5_console_stubs solo5_net_stubs)
+  (c_flags (:include cflags) -O2 -std=c99 -Wall -Werror)
```
Comment (avsm) [inline lib/dune]: I realise that the -Werror is existing behaviour, but it makes it really painful when doing production builds with a newer compiler that introduces some random warning.

Comment (hannesm) [reply lib/dune]: I share your concerns, but IMHO such a change (removing `-Werror`) deserves its own PR rather than being included (hidden) in a build-system-changing PR. Out of curiousity, is it possible in dune to specify specific `development` vs `release` c_flags (so there could be `-Werror` in development mode)?

**Comment by mato** [comment]
On Saturday, 25.05.2019 at 16:48, Anil Madhavapeddy wrote:
> avsm commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,18 @@
> +(library
> +  (name oS)
> +  (public_name mirage-solo5.internals)
> +  (modules lifecycle main mM oS solo5 time)
> +  (c_names alloc_pages_stubs atomic_stubs barrier_stubs checksum_stubs clock_stubs cstruct_stubs main mm_stubs solo5_block_stubs solo5_console_stubs solo5_net_stubs)
> +  (c_flags (:include cflags) -O2 -std=c99 -Wall -Werror)
> 
> I realise that the -Werror is existing behaviour, but it makes it really painful when doing production builds with a newer compiler that introduces some random warning.

Isn't that consistent with Dune behaviour for OCaml code, i.e. treating all
warnings as errors by default? Why deal with C code differently?

File: lib/main.ml (hunk 1 above)
Comment (TheLortex) [reply lib/main.ml]: @pascutto can correct me if I'm wrong, but it was because the use of `Lwt_sequence` is deprecated. Thanks to #42 we can remove that while rebasing.

**Comment by avsm** [comment]
Almost is, but not quite. Dune has profiles, and in the default dev mode it treats warnings as errors. But if `dune build --profile=release` or `dune build -p <package` is specified its in release mode, and they are just warnings.  We could do something similar by creating a `(profile)` in the dune-project that adds `-Werror` to CFLAGS only in dev builds

**Comment by emillon** [comment]
if `configurator` is used, it's possible to pass it the selected profile as argument, and the configure binary can output `-Werror` or not based on this piece of info.

**Comment by TheLortex** [comment]
@hannesm We can make a first PR that put solo5-specific modules in an `Os_solo5` module and update `mirage-(block|net|*)-solo5` accordingly. And then we can make `Os` an implementation of `mirage-os-shim`. What do you think ?
